operations/configuration_management/Write_Ansible_Code_for_CFS.md

```diff
@@ -11,22 +11,55 @@ Help with Ansible can be found in the external Ansible documentation:
 
-* Use image customization when possible to limit how many times a task is run and improve boot times. Configuration that is the same for all nodes of the same type will benefit from image customization. See the next section for how to target specific tasks for running only during image customization.
-* Import roles rather than playbooks. Each time a new playbook starts, Ansible automatically gathers facts for all the systems it is running against. This is not necessary more than once and can slow down Ansible execution.
+### Using Image Customization
+
+* Use image customization when possible to limit how many times a task is run and improve boot times. Configuration that is the same for all nodes of the same type will benefit from image customization.
+* Use different playbooks for image-customization and node-personalization.  Moving image customizations tasks to their own playbook can remove the need to evaluate that conditionals in a shared playbook, as well as ensuring that tasks are not accidentally running in both modes needlessly.
+* See [Target Ansible Tasks for Image Customization](Target_Ansible_Tasks_for_Image_Customization.md) for more information on writing for image customization.
+
+### Disable Fact Gathering
+
 * Turn off facts that are not needed in a playbook by setting `gather_facts: false`. If only a few facts are required, it is also possible to limit fact gathering by setting `gather_subset`. For more information on `gather_subset`, see the external [Ansible module setup](https://docs.ansible.com/ansible/latest/modules/setup_module.html) documentation.
+* Avoid importing playbooks in other playbooks.  This will trigger fact gathering for each imported playbook, potentially collecting the same information multiple times.
+
+### Reduce Wasted Time
+
+* Use `include_*` (dynamic re-use) to skip multiple tasks at once when using conditionals.  Ansible evaluates conditionals for every node in every task.  This includes when the conditional is applied to a block, or a role imported with `roles:` or the `import_role` task.  This is because these are static imports that are compiled at the beginning of the playbook, and the conditional is inherited by every task in the role or block.  Evaluating these conditionals for each task may only take a second or two, but across the hundreds of tasks that might be part of a playbook, this can add up to significant wasted time. Instead use dynamic imports with the `include_*` tasks.  Because these are evaluated at runtime, a conditional can skip the import of the role or tasks entirely, and is only evaluated once.  See the Ansible documentation on [Conditionals with re-use](https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html#conditionals-with-re-use) and [Re-using files and roles](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse.html#re-using-files-and-roles) for more information.
```

Comment:
```suggestion
* Use `include_*` (dynamic re-use) to skip multiple tasks at once when using conditionals. Ansible evaluates conditionals for every node in every task. This includes when the conditional is applied to a block, or a role imported with `roles:` or the `import_role` task.
This is because these are static imports that are compiled at the beginning of the playbook, and the conditional is inherited by every task in the role or block. Evaluating these conditionals for each task may only take a second or two, but across the hundreds of tasks that might be part of a playbook, this can add up to significant wasted time.
Instead use dynamic imports with the `include_*` tasks. Because these are evaluated at runtime, a conditional can skip the import of the role or tasks entirely, and is only evaluated once.
See the Ansible documentation on [Conditionals with re-use](https://docs.ansible.com/ansible/latest/user_guide/playbooks_conditionals.html#conditionals-with-re-use) and [Re-using files and roles](https://docs.ansible.com/ansible/latest/user_guide/playbooks_reuse.html#re-using-files-and-roles) for more information.
```